internal/metrics/middleware.go

```diff
@@ -0,0 +1,53 @@
+package metrics
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/lbryio/lbrytv/internal/errors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type key int
+
+const timerContextKey key = iota
+
+// Measure middleware starts a timer whenever a request is performed.
+// It should be added as first in the chain of middlewares.
+// Note that it doesn't catch any metrics by itself,
+// HTTP handlers are expected to add their own by calling AddObserver.
+func Measure() mux.MiddlewareFunc {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			t := StartTimer()
+			Logger.Log().Debugf("timer %p started", t)
+			rc := r.Clone(context.WithValue(r.Context(), timerContextKey, t))
+			next.ServeHTTP(w, rc)
+			t.Stop()
+			Logger.Log().Debugf("timer %p stopped (%.6fs)", t, t.Duration)
```

Comment:
if you put this in a deferred fn, will it still fire in case of a panic? just curious

Reply:
It will. Probably worth measuring panicked calls as well.